### PR TITLE
codeintel: Add associated upload and index records to UI

### LINKED ISSUE #21609: codeintel: Track upload/repository from which a dependency index job was created
Following up from action items listed in https://github.com/sourcegraph/sourcegraph/pull/21476.

### PR DESCRIPTION
**Pull request by efritz**
Expose associated upload and indexes in the UI. Fixes #21609.

Summary of code changes:

- Started fetching associatedIndex/associatedUpload when querying upload and index records
- Added and displayed `CodeIntelAssociatedUpload` and `CodeIntelAssociatedIndex` components
- Cleanup/drive-by:
  - Moved `<h3>` from timeline into where it's called
  - Moved `card` and `card-body` classes into `CodeIntelUploadMeta` and `CodeIntelIndexMeta` for consistency
  - Moved `d-flex flex-column align-items-center` classes out from `CodeIntelState`
  - Made types in shared code intel components more specific to its use of the types

Screen Shot 2021-06-04 at 12 28 25 PM

Screen Shot 2021-06-04 at 12 26 36 PM

### CODE REVIEW

File: enterprise/internal/codeintel/stores/dbstore/indexes.go [hunk 1]
```diff
@@ -375,6 +379,7 @@ var indexColumnsWithNullRank = []*sqlf.Query{
 	sqlf.Sprintf(`u.execution_logs`),
 	sqlf.Sprintf("NULL"),
 	sqlf.Sprintf(`u.local_steps`),
+	sqlf.Sprintf(`(SELECT id FROM lsif_uploads WHERE associated_index_id = u.id LIMIT 1) as associated_upload_id`),
```
Comment (efritz) [inline enterprise/internal/codeintel/stores/dbstore/indexes.go:382]: TODO:
- [ ] take max id
- [ ] ensure associated_index_id is indexed

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/upload.go [hunk 2]
```diff
@@ -43,6 +47,22 @@ func (r *UploadResolver) State() string {
...
 }
 
+func (r *UploadResolver) AssociatedIndex(ctx context.Context) (gql.LSIFIndexResolver, error) {
+	// TODO - why are a bunch of them zero?
+	if r.upload.AssociatedIndexID == nil || *r.upload.AssociatedIndexID == 0 {
+		return nil, nil
+	}
+
+	// TODO - bulk request everything in the layer
+	fmt.Printf("FETCHING INDEX %d FROM DATABASE\n", *r.upload.AssociatedIndexID)
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/upload.go:57]: TODO:
- [ ] add batch request layer here

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/index.go [hunk 3]
```diff
@@ -43,6 +47,21 @@ func (r *IndexResolver) State() string {
 	return state
 }
 
+func (r *IndexResolver) AssociatedUpload(ctx context.Context) (gql.LSIFUploadResolver, error) {
+	if r.index.AssociatedUploadID == nil {
+		return nil, nil
+	}
+
+	// TODO - bulk request everything in the layer
+	fmt.Printf("FETCHING UPLOAD %d FROM DATABASE\n", *r.index.AssociatedUploadID)
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/index.go:56]: TODO:
- [ ] add batch request layer here

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/upload.go [hunk 4]
```diff
@@ -43,6 +47,22 @@ func (r *UploadResolver) State() string {
 	return state
 }
 
+func (r *UploadResolver) AssociatedIndex(ctx context.Context) (gql.LSIFIndexResolver, error) {
+	// TODO - why are a bunch of them zero?
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/upload.go:51]: TODO:
- Troubleshoot zero insertions

File: client/web/src/enterprise/codeintel/detail/CodeIntelUploadPage.tsx [hunk 5]
```diff
@@ -146,6 +147,18 @@ export const CodeIntelUploadPage: FunctionComponent<CodeIntelUploadPageProps> =
...
+                    {uploadOrError.associatedIndex && uploadOrError.projectRoot?.repository && (
+                        <div>
+                            <Link
+                                to={`/${uploadOrError.projectRoot.repository.name}/-/settings/code-intelligence/indexes/${uploadOrError.associatedIndex.id}`}
+                            >
+                                Associated index
+                            </Link>{' '}
+                            is {uploadOrError.associatedIndex.state}
+                        </div>
+                    )}
```
Comment (efritz) [inline client/web/src/enterprise/codeintel/detail/CodeIntelUploadPage.tsx:161]: TODO:
- [ ] restyle

File: client/web/src/enterprise/codeintel/detail/CodeIntelIndexPage.tsx [hunk 6]
```diff
@@ -124,6 +125,18 @@ export const CodeIntelIndexPage: FunctionComponent<CodeIntelIndexPageProps> = ({
...
+                    {indexOrError.associatedUpload && indexOrError.projectRoot?.repository && (
+                        <div>
+                            <Link
+                                to={`/${indexOrError.projectRoot.repository.name}/-/settings/code-intelligence/uploads/${indexOrError.associatedUpload.id}`}
+                            >
+                                Associated upload
+                            </Link>{' '}
+                            is {indexOrError.associatedUpload.state}
+                        </div>
+                    )}
```
Comment (efritz) [inline client/web/src/enterprise/codeintel/detail/CodeIntelIndexPage.tsx:139]: TODO:
- [ ] restyle

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/index.go [hunk 7]
```diff
@@ -7,18 +7,27 @@ import (
...
 
 type IndexResolver struct {
 	index            store.Index
+	resolver         resolvers.Resolver
+	prefetcher       *Prefetcher
 	locationResolver *CachedLocationResolver
 }
 
-func NewIndexResolver(index store.Index, locationResolver *CachedLocationResolver) gql.LSIFIndexResolver {
+func NewIndexResolver(index store.Index, resolver resolvers.Resolver, prefetcher *Prefetcher, locationResolver *CachedLocationResolver) gql.LSIFIndexResolver {
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/index.go:22]: resolver not actually necessary now

File: client/web/src/enterprise/codeintel/list/CodeIntelUploadNode.tsx [hunk 8]
```diff
@@ -40,7 +40,7 @@ export const CodeIntelUploadNode: FunctionComponent<CodeIntelUploadNodeProps> =
         </div>
 
         <span className="d-none d-md-inline codeintel-upload-node__state">
-            <CodeIntelState node={node} />
+            <CodeIntelState node={node} className="d-flex flex-column align-items-center " />
```
Comment (efritz) [inline client/web/src/enterprise/codeintel/list/CodeIntelUploadNode.tsx:43]:
```suggestion
            <CodeIntelState node={node} className="d-flex flex-column align-items-center" />
```

File: client/web/src/enterprise/codeintel/detail/CodeIntelAssociatedUpload.tsx [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+                    <span className="d-none d-md-inline codeintel-associated-upload__state">
+                        <CodeIntelState
+                            node={node.associatedUpload}
+                            className="d-flex flex-column align-items-center"
+                        />
+                    </span>
+                    <span>
+                        <Link
+                            to={`/${node.projectRoot.repository.name}/-/settings/code-intelligence/uploads/${node.associatedUpload.id}`}
```
Comment (eseliger) [inline client/web/src/enterprise/codeintel/detail/CodeIntelAssociatedUpload.tsx:49]: Whoa what an URL

**Review by eseliger — APPROVED** [review]
Nice work 🌟

File: client/web/src/enterprise/codeintel/detail/CodeIntelAssociatedUpload.tsx (hunk 9 above)
Comment (efritz) [reply client/web/src/enterprise/codeintel/detail/CodeIntelAssociatedUpload.tsx:49]: "an you are elle"

Comment (eseliger) [reply client/web/src/enterprise/codeintel/detail/CodeIntelAssociatedUpload.tsx:49]: 1 ell of a URL

**Pull request merged**